tsconfig.base.json

```diff
@@ -0,0 +1,27 @@
+{
+  "compilerOptions": {
+    "target": "ES2015",
+    "module": "esnext",
+    "moduleResolution": "node",
+    "lib": ["ES2015", "dom"],
+    "sourceMap": true,
+    "strict": true,
+    "strictFunctionTypes": false,
+    "forceConsistentCasingInFileNames": true,
+    "esModuleInterop": true,
```

Comment:
I don't think this should be included in the defaults since it can cause problems downstream if not all consumers are also using it. IIRC @layershifter may have more background here.

Reply:
As I remember it will require `esModuleInterop: true` on customer's side, too. Not all our customers have it enabled.

I failed to find a PR that have been done a long time ago, but at least a commit is there (https://github.com/microsoft/fluent-ui-react/commit/75c139d0c1ed4f2ef0b8f886ff8df53bad74ac56).

Reply:
> I failed to find a PR that have been done a long time ago, but at least a commit is there (microsoft/fluent-ui-react@75c139d).

unfortunately that's not really helpful, no reasoning etc :) 

---

Overall I'm not aware of any implications to customers nor about a requirement to have this enabled as well in consumers setup. 

esModuleInterop helps to consume non es2015 module code with JS module syntax and normalizes ambient definitions consumption. That's it (no changes to export declaration emit for consumers, nor from runtime perspective). Unless you can provide explicit issue I'm gonna keep it as is. thx
- [official docs](https://www.typescriptlang.org/tsconfig#esModuleInterop)
- [SO](https://stackoverflow.com/questions/56238356/understanding-esmoduleinterop-in-tsconfig-file)

Reply:
I created a sample project, https://github.com/layershifter/tsc-test

Where:
- `host` is an application that has `esModuleInterop: true` and builds type definitions
- `consumer` is an application that has `esModuleInterop: false` (i.e. default value) and consumes `host`

To run it:
- clone
- `yarn`
- `yarn build`

You will a build failure:

```
$ tsc
../host/dist/index.d.ts:1:8 - error TS1259: Module '"C:/Users/olfedias/WebstormProjects/tsc-test/node_modules/@types/react/index"' can only be default-imported using the 'esModuleInterop' flag

1 import React from 'react';
         ~~~~~

  ../../node_modules/@types/react/index.d.ts:69:1
    69 export = React;
       ~~~~~~~~~~~~~~~
    This module is declared with using 'export =', and can only be used with a default import when using the 'esModuleInterop' flag.
```

`skipLibCheck: true` can workaround it, but it also defaults to `false`. If we will enable `esModuleInterop` produced artifacts cannot be consumed by customers with default TypeScript configs.

Reply:
Thanks for digging into it. I forgot about this completely (I guess sideffect for using skip lib check for couple of years everywhere)  

> I run into that in our codebase quite recently as well  🗡️  https://github.com/microsoft/fluentui/pull/17033/commits/74f5de4102c5618fe2bd84e68775fbcb786a837c )

This leads us back to fundamental question: What are requirements for fluent convergence regarding ambient declarations output ?

All in all this feels like right change not just to us but for every customer.

Why?
- skipLibCheck is recommended to be set to `true` in official docs (it's not true by default to not introduce breaking changes, which are introduced anyway even in feature bumps)
- if customers enable this setting they will get type checking speed boost "for free"

Reply:
IMO, our packages should work with any TS config including default options as consumers can have different motivation for their `compilerOptions`. For example, `esModuleInterop` is not enabled in Teams, this change will require them either to enable `skipLibCheck` everywhere (_they might not want to do this everywhere_) or `esModuleInterop`.

---

Personally I also don't see any value from enabling `esModuleInterop` (for components' packages) as in general we should not use CommonJS dependencies (they are impacting tree-shaking). We have React's packages as CommonJS dependencies, but they are going to have ESM builds (https://github.com/facebook/react/issues/11503) and star imports (`import * as React`) are recommended currently anyway:

> The default imports will keep working in React 17, but in the longer term we encourage moving away from them.
https://ru.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#removing-unused-react-imports

Reply:
> IMO, our packages should work with any TS config 

This is way to ambitious statement :) do we support `moduleResolution: "classic" ` etc ?  ...
Consumer of a package shouldn't care nor be limited by library internal tsconfig used.

> require them either to enable skipLibCheck everywhere 

Thats good for them (see my previous comment), TS exec speed benefits. 

> as in general we should not use CommonJS dependencies 

I agree, although esModuleInterop won't help you with that much :) we'd need special lint rule or even better explicit deps validator 

> (import * as React) are recommended currently anyway:

It's not recommended way rather than "valid way that will not get replaced when running codemod". I already touched this point that we should rather use explicit named imports. anyways this is out off topic in terms of this PR.